src/world.rs

```diff
@@ -0,0 +1,311 @@
+extern crate alloc;
+use crate::display::Display;
+use core::fmt::Write;
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use pc_keyboard::{DecodedKey, KeyCode};
+
+#[derive(PartialEq, Debug)]
+pub enum GameState {
+    Live,
+    GameOver,
+}
+
+pub trait Game {
+    fn on_keypress(&mut self, key: DecodedKey);
+    fn reset(&mut self, width: usize, height: usize);
+
+    // meaning of return value:
+    //    true  => game over
+    //    false => continue
+    fn step(&mut self) -> GameState;
+    fn draw(&mut self, display: &mut Display);
+}
+
+#[derive(Clone, Debug)]
+enum State {
+    Welcome((bool, usize)), // (dirty, selected_game)
+    Running(usize),
+    GameOver(bool), // dirty
+}
+
+unsafe impl Send for World {}
+
+pub struct World {
+    games: Vec<Box<dyn Game>>,
+    game_names: Vec<&'static str>,
+    state: State,
+    width: usize,
+    height: usize,
+}
+
+impl World {
+    pub fn new(width: usize, height: usize) -> Self {
+        Self {
+            games: Vec::new(),
+            game_names: Vec::new(),
+            state: State::Welcome((true, 0)),
+            width,
+            height,
+        }
+    }
+
+    pub fn add_game(&mut self, mut game: Box<dyn Game>, name: &'static str) {
+        game.reset(self.width, self.height);
+        self.games.push(game);
+        self.game_names.push(name);
+    }
+
+    pub fn on_keypress(&mut self, key: DecodedKey, display: &mut Display) {
+        match self.state {
+            State::Running(i) if key == DecodedKey::Unicode('r') => {
+                self.state = State::Welcome((true, 0));
+            }
+            State::Running(i) => {
+                let w = &mut self.games[i];
+                w.on_keypress(key);
+            }
+            State::Welcome((_, selected_game)) => {
+                match key {
+                    DecodedKey::Unicode('q') => {
+                        // TODO: how to quit in no_std?
```

Comment:
I guess we can reboot somehow I think that is covered in the blog in the article about testing.

Reply:
According to https://os.phil-opp.com/testing/, this is much more complicated than I thought.
```txt
The clean solution to this would be to implement a proper way to shutdown our OS.
Unfortunately this is relatively complex, because it requires implementing support for either
the APM or ACPI power management standard.
```

Just leave it as this, maybe fix it later